modules/files.nix

```diff
@@ -197,6 +197,37 @@ in
         ''
     );
 
+    home.activation.checkFilesChanged = dag.entryBefore ["linkGeneration"] (
+      ''
+        declare -A changed_files
+      '' + concatStrings (
+        mapAttrsToList (n: v: ''
+
+        changed_files["${v.target}"]=0
+        cmp --quiet \
+        "${v.source}" \
+        "${v.target}" \
+        || changed_files["${v.target}"]=1
+        '')
+        (filterAttrs (n: v: v.onChange != null)
+        cfg)
+      )
+    );
+
+    home.activation.onFilesChange = dag.entryAfter [ "linkGeneration" ] (
+      concatStrings (
+        # TODO: find nicer way to escape variable substitution
+        mapAttrsToList (n: v: ''
+
+          if [ ${"$\{changed_files"}["${v.target}"]} -eq 1 ]; then
```

Comment:
Hmm, maybe you want

    if [[ ''${changed_files["${v.target}"]} -eq 1 ]]; then

?